### ISSUE TITLE
Less boilerplate way to serialize/deserialize on native

### ISSUE THREAD

**Issue by luca992** [Kotlin/kotlinx.serialization#237]
I have figured out how to serialize and deserialize json on native... But the code is pretty verbose.

On native is there an easier way to accomplish this?

for example to deserialize the following on native:

    {
      "data": {
        "username": "John Snow",
         "user_id": 44
      }
    }

_

    @Serializable
    data class ApiSessionResponseContainer(
            val data: ApiSessionResponse) {
    
        @Serializer(forClass = ApiSessionResponseContainer::class)
        companion object : KSerializer<ApiSessionResponseContainer> {
            override fun serialize(output: Encoder, obj: ApiSessionResponseContainer) {
            }
            override fun deserialize(input: Decoder): ApiSessionResponseContainer {
                val input = input.beginStructure(descriptor)
                var data : ApiSessionResponse? = null
                mainLoop@ while (true) {
                    when (input.decodeElementIndex(descriptor)) {
                        READ_DONE -> {
                            break@mainLoop
                        }
                        0 -> {
                            val loader = input.context!!.get(ApiSessionResponse::class)!!
                            data = input.decodeSerializableElement<ApiSessionResponse>(descriptor, 0, loader)
                        }
                        else -> throw SerializationException("Invalid index")
                    }
                }
                input.endStructure(descriptor)
                return ApiSessionResponseContainer(data!!)
            }
    
            override val descriptor = object : SerialDescriptor {
                override val name = "co.example.data.api.response.ApiSessionResponseContainer"
                override val kind: SerialKind = StructureKind.CLASS
                override fun getElementName(index: Int) = when(index) {
                    0 -> "data"
                    else -> ""
                }
                override fun getElementIndex(name: String) = when(name) {
                    "data" -> 0
                    else -> -1
                }
                override fun isElementOptional(index: Int): Boolean = false
            }
        }
    }

_

    @Serializable
    data class ApiSessionResponse(
            @SerialName("username") val username: String,
            @SerialName("user_id") val userId: Long) {
    
        @Serializer(forClass = ApiSessionResponse::class)
        companion object : KSerializer<ApiSessionResponse> {
            override fun serialize(output: Encoder, obj: ApiSessionResponse) {
            }
            override fun deserialize(input: Decoder): ApiSessionResponse {
                val input = input.beginStructure(descriptor)
                val data = mutableMapOf<String,Any>()
                mainLoop@ while (true) {
                    when (input.decodeElementIndex(descriptor)) {
                        READ_DONE -> {
                            break@mainLoop
                        }
                        0 -> {
                            data["username"] = input.decodeStringElement(descriptor, 0)
                        }
                        1 -> {
                            data["userId"] = input.decodeLongElement(descriptor, 1)
                        }
                        else -> throw SerializationException("Invalid index")
                    }
                }
                input.endStructure(descriptor)
                return ApiSessionResponse(data["username"] as String ,data["userId"] as Long)
            }
            override val descriptor = object : SerialDescriptor {
                override val name = "co.example.data.api.response.ApiSessionResponse"
                override val kind: SerialKind = StructureKind.CLASS
                override fun getElementName(index: Int) = when(index) {
                    0 -> "username"
                    1 -> "user_id"
                    else -> ""
                }
                override fun getElementIndex(name: String) = when(name) {
                    "username" -> 0
                    "user_id" -> 1
                    else -> -1
                }
                override fun isElementOptional(index: Int): Boolean = false
            }
        }
    }

then deserializing:

    val module = object : SerialModule {
        override fun registerIn(context: MutableSerialContext) {
        context.registerSerializer(ApiSessionResponseContainer::class, ApiSessionResponseContainer)
        context.registerSerializer(ApiSessionResponse::class, ApiSessionResponse)
        }
    }
    val json = JSON(unquoted = false).apply { install(module) }
    val body = """{"data":{"username":"John Snow","user_id":44}}"""
    val apiSessionResponseContainer = json.parse<ApiSessionResponseContainer>(body)

**Comment by sandwwraith** [comment]
Yes, this is exactly the code required to serialize and deserialize classes. Writing it by hand is really quite verbose, so there is a compiler plugin for that purpose, have you tried to enable it (it is already available on native with some restrictions)? One caveat: You'll have to pass serializer explicitly, like this: 
`val apiSessionResponseContainer = json.parse(ApiSessionResponseContainer.serializer(), body)`

**Comment by luca992** [comment]
Huh, I did not know the plugin worked for native. I see you added that to the readme. That would have been nice to know before haha.
